compiler/rustc_session/src/config.rs

```diff
@@ -1528,23 +1527,26 @@ fn parse_libs(
         .map(|s| {
             // Parse string of the form "[KIND=]lib[:new_name]",
             // where KIND is one of "dylib", "framework", "static".
-            let mut parts = s.splitn(2, '=');
-            let kind = parts.next().unwrap();
-            let (name, kind) = match (parts.next(), kind) {
-                (None, name) => (name, NativeLibKind::Unspecified),
-                (Some(name), "dylib") => (name, NativeLibKind::Dylib),
-                (Some(name), "framework") => (name, NativeLibKind::Framework),
-                (Some(name), "static") => (name, NativeLibKind::StaticBundle),
-                (Some(name), "static-nobundle") => (name, NativeLibKind::StaticNoBundle),
-                (_, s) => {
-                    early_error(
-                        error_format,
-                        &format!(
-                            "unknown library kind `{}`, expected \
-                             one of dylib, framework, or static",
-                            s
-                        ),
-                    );
```

Comment:
Just writing my reasoning down for posterity:

The previous version is more compact, so I am *almost* inclined to say that it was better. However, the new version doesn't have `.unwrap()`, and, given a couple of instances where have impossible cases in other parts of this PR, I think a tad more verbosity is a fine price to pay for making illegal states unrepresentable here. 

That sad, would flattening the two matches into one work better here?

```rust
match s.split_once('=') {
  Some(("dylib", name)) => ()
}
```

?

Reply:
>Instead, what do you think about adding impl TryFrom<String> for NativeLibKind?

Not `TryFrom`, but `impl FromStr for NativeLibKind` would work nicely. But I think the added benefit here is marginal, if any (there's drawback of more code churn), so I suggest leaving the code as is then!